### PR TITLE
Check len(freeMessages) in DrainContext

### LINKED ISSUE #18: Sometimes messages are not printed even when you call `logger.Drain`
`logger.Drain` calls `logger.DrainContext`, which works by checking the msg queue length and sleeping 10ms if there's something in it.

But there's a race between `logWriter` reading `messages`, `DrainContext` checking its length, and `logWriter` actually printing the message.  It's entirely possible that DrainContext can see no messages in the queue, and return, and its caller call `os.Exit`, before `logWriter` prints the message.  So you don't get the message even though you called `logger.Drain` with no timeout.

### PR DESCRIPTION
**Pull request by theclapp**
Also create DrainWithTimeout.

Fixes #18.

### CODE REVIEW

File: logger/logger_writer.go [hunk 1]
```diff
@@ -311,14 +311,22 @@ func Close(ctx context.Context) error {
 // Pending messages may never run out if another goroutine is constantly
 // writing.
 func DrainContext(ctx context.Context) error {
-	for ctx.Err() == nil && len(messages) > 0 {
+	for ctx.Err() == nil && (len(messages) > 0 || len(freeMessages) < cap(freeMessages)) {
 		innerCtx, cancel := context.WithTimeout(ctx, 10*time.Millisecond)
-		<-innerCtx.Done() // Wait for 10ms and check len(messages) again
+		<-innerCtx.Done() // Wait for 10ms and check the queues again
```
Comment (cleroux) [inline logger/logger_writer.go:316]: I'm not seeing how this is different than just calling `time.Sleep(10*time.Millisecond)`.

I guess a good reason to have `innerCtx` would be to cancel the timeout when `ctx` is cancelled.  But that's not implemented here.  And I guess 10ms is short enough that we don't care.

**Review by cleroux — APPROVED** [review]
I guess it makes sense to make sure freeMessages is full again.  Did you actually observe a message not being drained because of this?  If so, amazing find.

File: logger/logger_writer.go (hunk 1 above)
Comment (theclapp) [reply logger/logger_writer.go:316]: > I'm not seeing how this is different than just calling time.Sleep(10*time.Millisecond).

The outer context `ctx` might timeout (or be cancelled) sooner than 10ms.

> I guess a good reason to have innerCtx would be to cancel the timeout when ctx is cancelled. But that's not implemented here.

It is.  Context cancellation is transitive.  "When a Context is canceled, all Contexts derived from it are also canceled."

**Comment by theclapp** [comment]
> Did you actually observe a message not being drained because of this? If so, amazing find.

Yes, first in fra1 and then locally.  It was consistently reproducible.

File: logger/logger_writer.go (hunk 1 above)
Comment (cleroux) [reply logger/logger_writer.go:316]: oh shoot, I missed that detail.  Thanks for correcting me.

**Pull request merged**